src/main/cpp/utils/StateManager.cpp

```diff
@@ -0,0 +1,52 @@
+#include "utils/StateManager.h"
+
+template <typename StateType>
+StateManager<StateType>::StateManager(const std::function<bool()> Interrupt,
+                                      StateType DefaultState,
+                                      frc2::CommandPtr DefaultEntrypoint)
+    : m_currentState(DefaultState) {}
+
+template <typename StateType>
+void StateManager<StateType>::RegisterState(StateType State,
+                                            frc2::CommandPtr Entrypoint,
+                                            units::second_t Timeout) {
+  m_cmds[State] = std::move(Entrypoint).WithTimeout(Timeout);
+}
+
+template <typename StateType>
+void StateManager<StateType>::RegisterState(StateType State,
+                                            frc2::CommandPtr Entrypoint) {
+  m_cmds[State] = std::move(Entrypoint);
+}
+
+template <typename StateType>
+frc2::CommandPtr StateManager<StateType>::GetStateCommand(StateType State) {
+  return m_cmds[State];
+}
+
+template <typename StateType>
+frc2::CommandPtr StateManager<StateType>::SetState(StateType newState) {
+  return frc2::InstantCommand([this, newState] {
+           m_currentState = newState;
+           RunState();
+         })
+      .ToPtr();
+}
+
+template <typename StateType>
+frc2::DeferredCommand StateManager<StateType>::RunStateDeferred() {
+  return frc2::DeferredCommand([this] { RunState(); }, {this});
+}
+
+template <typename StateType>
+frc2::CommandPtr StateManager<StateType>::RunState() {
+  return m_cmds[m_currentState];
+}
+
+template <typename StateType>
+void StateManager<StateType>::SetDesiredState() {
```

Comment:
My bad on the naming convention, but this seems to be running the command rather than just setting it